src/Pos/Explorer/BListener.hs

```diff
@@ -0,0 +1,481 @@
+        putKeyBlocks keyBlocks = DB.PutEpochBlocks key uniqueBlocks
+          where
+            key           = keyBlocks ^. _1
+            blocks        = keyBlocks ^. _2
+            uniqueBlocks  = ordNub blocks
+
+
+-- For @PageBlocks@
+onApplyPageBlocksExplorer
+    :: forall ssc m .
+    MonadBListenerT m ssc
+    => OldestFirst NE (Blund ssc) 
+    -> m SomeBatchOp
+onApplyPageBlocksExplorer blunds = do
+
+    -- Get existing @HeaderHash@es from the keys so we can merge them.
+    existingBlocks <- getExistingBlocks $ M.keys newBlocks
+
+    -- Merge the new and the old
+    let mergedBlocksMap = M.unionWith (<>) existingBlocks newBlocks
+
+    -- Create database operation
+    let mergedBlocks = putKeysBlocks mergedBlocksMap
+
+    -- In the end make sure we place this under @SomeBatchOp@ in order to
+    -- preserve atomicity.
+    pure $ SomeBatchOp mergedBlocks
+  where
+
+    newBlocks :: M.Map Page [HeaderHash]
+    newBlocks = pageBlocksMap blocksNE
+
+    blocksNE :: NE (Block ssc)  
+    blocksNE = fst <$> getOldestFirst blunds
+
+    -- Get exisiting key blocks paired with the key.
+    getExistingBlocks
+        :: (MonadDBRead m) 
+        => [Page]
+        -> m (M.Map Page [HeaderHash])
+    getExistingBlocks keys = do
+        keyBlocks <- sequence [ getExistingKeyBlocks key | key <- keys ]
+        pure $ M.unions keyBlocks
+      where
+        -- Get exisiting key blocks paired with the key. If there are no
+        -- saved blocks on the key return an empty list.
+        getExistingKeyBlocks 
+            :: (MonadDBRead m)
+            => Page 
+            -> m (M.Map Page [HeaderHash])
+        getExistingKeyBlocks key = do
+            keyBlocks        <- DB.getPageBlocks key
+            let mKeyBlocks = fromMaybe [] keyBlocks
+            pure $ M.singleton key mKeyBlocks
+
+    -- For each (k, [v]) pair, create a database operation.
+    putKeysBlocks 
+        :: M.Map Page [HeaderHash]
+        -> [DB.ExplorerOp]
+    putKeysBlocks keysBlocks = putKeyBlocks <$> M.toList keysBlocks
+      where
+        putKeyBlocks 
+            :: (Page, [HeaderHash])
+            -> DB.ExplorerOp
+        putKeyBlocks keyBlocks = DB.PutPageBlocks key uniqueBlocks
+          where
+            key           = keyBlocks ^. _1
+            blocks        = keyBlocks ^. _2
+            uniqueBlocks  = ordNub blocks
+    
+
+----------------------------------------------------------------------------
+-- Rollback
+----------------------------------------------------------------------------
+
+-- For @EpochBlocks@
+onRollbackEpochBlocksExplorer
+    :: forall ssc m .
+    MonadBListenerT m ssc
+    => NewestFirst NE (Blund ssc) -> m SomeBatchOp
+onRollbackEpochBlocksExplorer blunds = do
+    
+    -- Get existing @HeaderHash@es from the keys so we can merge them.
+    existingBlocks <- getExistingBlocks keys
+
+    -- Diffrence between the new and the old
+    let mergedBlocksMap = rollbackedBlocks keys newBlocks existingBlocks
```

Comment:
I see `rollbackedBlocks keys' oldMap' newMap'`, but here you pass `newBlocks` before `existingBlocks`. Is it by mistake?